core/src/main/java/hudson/cli/DisablePluginCommand.java

```diff
@@ -0,0 +1,135 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.cli;
+
+import hudson.Extension;
+import hudson.PluginManager;
+import hudson.PluginWrapper;
+import jenkins.model.Jenkins;
+import org.kohsuke.args4j.Argument;
+import org.kohsuke.args4j.Option;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Disable one or more installed plugins. If the plugin has an enabled dependant then it can't be disabled, the process
+ * continues, but the exit code is 16, instead of 0. Disabling an already disabled plugin does nothing. It only restart
+ * if, at least, one plugin has been disabled and the restart option was set.
+ *
+ * @since 2.136
+ */
+@Extension
+public class DisablePluginCommand extends CLICommand {
+
+    @Argument(metaVar = "plugins", required = true, usage = "Plugins to be disabled.")
+    private List<String> pluginNames;
+
+    @Option(name = "-restart", usage = "Restart Jenkins after disabling plugins.")
+    private boolean restart;
+
+    @Override
+    public String getShortDescription() {
+        return Messages.DisablePluginCommand_ShortDescription();
+    }
+
+    @Override
+    protected void printUsageSummary(PrintStream stderr) {
+        stderr.println(Messages.DisablePluginCommand_PrintUsageSummary());
+    }
+
+    @Override
+    protected int run() throws Exception {
+        Jenkins jenkins = Jenkins.get();
+        jenkins.checkPermission(Jenkins.ADMINISTER);
+
+        PluginManager manager = jenkins.getPluginManager();
+        int result = 0;
+        boolean somePluginDisabled = false; // to ensure we restart when it needs to
+        for (String pluginName : pluginNames) {
+            switch (disablePlugin(manager, pluginName)) {
+                case DISABLED:
+                    if(!somePluginDisabled) somePluginDisabled = true;
+                    break;
+                case NOT_DISABLED_DEPENDANTS:
+                    if(result != 16) result = 16; // Custom error code to indicate one plugin, at least, can't be disabled
+            }
+        }
+
+        if (restart && somePluginDisabled) {
+            jenkins.safeRestart();
```

Comment:
Is that an asynchronous execution? I mean, will the CLI wait for the restart?

Reply:
It doesn't wait. The restart will begin when there aren't jobs running